packages/client/src/schema/filters.ts

```diff
@@ -114,12 +115,27 @@ export type Filter<T> = T extends Record<string, any>
     : BaseApiFilter<T> | NestedApiFilter<T>
   : PropertyFilter<T>;
 
-export function cleanFilter(filter?: FilterExpression) {
-  if (!filter) return undefined;
+export function cleanFilter(filter?: FilterExpression | FilterPredicate): FilterExpression | undefined {
+  if (!isDefined(filter)) return undefined;
 
-  const values = Object.values(filter)
-    .filter(Boolean)
-    .filter((value) => (Array.isArray(value) ? value.length > 0 : true));
+  const values = Object.fromEntries(
+    Object.entries(filter).reduce((acc, [key, value]) => {
+      // Remove null and undefined values
+      if (!isDefined(value)) return acc;
 
-  return values.length > 0 ? filter : undefined;
+      if (Array.isArray(value)) {
+        // Remove empty objects from arrays
+        const clean = value.map((item) => cleanFilter(item)).filter((item) => isDefined(item));
```

Comment:
This is the real change, for arrays we now iterate and clean the values from undefined and nulls. The rest of diff is changing from `filter` to `reduce` to do the same checks.